packages/api/src/routes/medical/document.ts

```diff
@@ -108,30 +121,145 @@ router.post(
+  const conversionType = type ? docConversionTypeSchema.parse(type) : undefined;
+
+  if ((typeof fileName !== "string" || fileName.indexOf(cxId) !== -1) && !Config.isSandbox()) {
+    throw new ForbiddenError();
+  }
+
+  const url = await downloadDocument({ fileName, conversionType });
+  return url;
+}
+
+// TODO: Redirect this endpoint to the new one (download-url)
+/** ---------------------------------------------------------------------------
+ * GET /document/downloadUrl
  * Fetches the document from S3 and sends a presigned URL
+ * @deprecated Use the GET /download-url endpoint instead.
  *
  * @param req.query.fileName The file name of the document in s3.
  * @param req.query.conversionType The doc type to convert to.
  * @return presigned url
  */
 router.get(
   "/downloadUrl",
+  asyncHandler(async (req: Request, res: Response) => {
+    const url = getDownloadUrl(req);
+    return res.status(MOVED_PERMANENTLY).json({ url });
+  })
+);
+
+/** ---------------------------------------------------------------------------
+ * GET /document/download-url
+ *
+ * Fetches the document from S3 and sends a presigned URL
+ *
+ * @param req.query.fileName The file name of the document in s3.
+ * @param req.query.conversionType The doc type to convert to.
+ * @return presigned url
+ */
+router.get(
+  "/download-url",
+  asyncHandler(async (req: Request, res: Response) => {
+    const url = getDownloadUrl(req);
+    return res.status(OK).json({ url });
+  })
+);
+
+/**
+ * POST /document/upload
+ *
+ * Uploads a medical document and creates a Document Reference for that file on the FHIR server.
+ *
+ * @param patientId - The patientId of the patient.
+ *
+ * @return documentReferenceID for the uploaded file.
+ * Refer to Metriport Documentation for more details:
+ * https://docs.metriport.com/medical-api/api-reference/document/post-upload-url
+ */
+router.post(
+  "/upload",
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getCxIdOrFail(req);
-    const fileName = getFromQueryOrFail("fileName", req);
-    const fileHasCxId = fileName.includes(cxId);
-    const type = getFrom("query").optional("conversionType", req);
-    const conversionType = type ? docConversionTypeSchema.parse(type) : undefined;
+    const patientId = getFromQueryOrFail("patientId", req);
+    const fileId = uuidv7();
+    const s3FileName = createS3FileName(cxId, patientId, fileId);
+    const docRefId = uuidv7();
+    const s3Key = s3FileName + "_upload_" + docRefId;
 
-    if (!fileHasCxId && !Config.isSandbox()) throw new ForbiddenError();
+    const organization = await getOrganization({ cxId });
+    if (!organization) throw new ForbiddenError(`Organization not found for CX ${cxId}`);
 
-    const url = await downloadDocument({ fileName, conversionType });
+    const docRefDraft = req.body.docRef;
+    const fileContent = req.body.fileContent;
+    console.log("File Content", JSON.stringify(fileContent));
```

Comment:
Why are we getting the contents on our API? All this time we've been pushing to use signed URL so the cx gets the URL and they upload the file from their end w/o going through our API.

Reply:
So we still want to return the URL on this endpoiint then? Was using this `put` function just a part of the testing process? I'm confused..

Reply:
If we were to upload the file through our API we wouldn't need the pre-signed URL, would just `s3.putObject()`, right?

Remember [the whole conversation about it](https://metriport.slack.com/archives/C04DMKE9DME/p1697085761272189)?

Mainly:
1. API GW has a limit of 10MB for the payload;
2. We didn't want to be processing large files, which could potentially take down our API;
3. S3 does the plumbing of processing a file so we don't have to.